scripts/library.sh

```diff
@@ -475,21 +475,51 @@ function get_canonical_path() {
   echo "$(cd ${path%/*} && echo $PWD/${path##*/})"
 }
 
-# Return the base url we use to build the actual knative yaml sources.
-function get_knative_base_yaml_source() {
-  local knative_base_yaml_source="https://storage.googleapis.com/knative-nightly/@/latest"
+# Returns latest yaml source for the given knative project.
+# Parameters: $1 - repository name of the given project
+#             $2 - short name of the yaml source file
+function get_latest_knative_yaml_source() {
   local branch_name=""
+  local repo_name=$1
+  local source_name=$2
   # Get the branch name from Prow's env var, see https://github.com/kubernetes/test-infra/blob/master/prow/jobs.md.
   # Otherwise, try getting the current branch from git.
   (( IS_PROW )) && branch_name="${PULL_BASE_REF:-}"
   [[ -z "${branch_name}" ]] && branch_name="$(git rev-parse --abbrev-ref HEAD)"
-  # If it's a release branch, base URL should point to a specific version.
+  # If it's a release branch, the yaml source URL should point to a specific version.
   if [[ ${branch_name} =~ ^release-[0-9\.]+$ ]]; then
     # Get the latest tag name for the current branch, which is likely formatted as v0.5.0
     local tag_name="$(git describe --tags --abbrev=0)"
-    knative_base_yaml_source="https://storage.googleapis.com/knative-releases/@/previous/${tag_name}"
+    # The other repo might not have this tag, so we need to find the first available tag before it.
+    echo "$(find_latest_release_yaml_source ${tag_name} ${repo_name} ${source_name})"
+  # If it's not a release branch, the yaml source URL should be nightly build.
+  else
+    echo "https://storage.googleapis.com/knative-nightly/${repo_name}/latest/${source_name}.yaml"
   fi
-  echo "${knative_base_yaml_source}"
+}
+
+# Returns latest release yaml source for the given knative project.
+# Parameters: $1 - git tag name of the current branch, e.g. v0.5.1
+#             $2 - repository name of the given project
+#             $3 - short name of the yaml source file
+function find_latest_release_yaml_source() {
+  local tag_name=$1
+  local repo_name=$2
+  local source_name=$3
+  local major_minor="$(echo ${tag_name} | cut -d. -f-2)"
+  local revision="$(echo ${tag_name} | cut -d. -f3)"
+  for (( ; revision>=0; revision-- ))
```

Comment:
Drop the loop; list the dir, filter the subdirs by major/minor version, then reverse sort the list and use the first entry.

Reply:
Thanks for the suggestion!
Apparently I went the wrong direction, as I didn't know we can use `gsutil` in our scripts.
I have simplified the logic by using `gsutil ls` and `tail` and I have tested it's working.